File: helfomat-api/src/main/java/de/helfenkannjeder/helfomat/api/organization/OrganizationAssembler.java
package de.helfenkannjeder.helfomat.api.organization;

import com.google.common.base.Objects;
import de.helfenkannjeder.helfomat.api.question.AnsweredQuestionDto;
import de.helfenkannjeder.helfomat.core.organization.Address;
import de.helfenkannjeder.helfomat.core.organization.AttendanceTime;
import de.helfenkannjeder.helfomat.core.organization.ContactPerson;
import de.helfenkannjeder.helfomat.core.organization.Group;
import de.helfenkannjeder.helfomat.core.organization.Organization;
import de.helfenkannjeder.helfomat.core.organization.OrganizationId;
import de.helfenkannjeder.helfomat.core.organization.QuestionAnswer;
import de.helfenkannjeder.helfomat.core.organization.ScoredOrganization;
import de.helfenkannjeder.helfomat.core.organization.Volunteer;
import de.helfenkannjeder.helfomat.core.picture.PictureId;
import de.helfenkannjeder.helfomat.core.question.Question;
import de.helfenkannjeder.helfomat.core.question.QuestionId;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Valentin Zickner
 */
public class OrganizationAssembler {

    static List<OrganizationDto> toScoredOrganizationDtos(List<ScoredOrganization> organizations) {
        return organizations
            .stream()
            .map(OrganizationAssembler::toOrganizationDto)
            .collect(Collectors.toList());
    }

    private static OrganizationDto toOrganizationDto(ScoredOrganization scoredOrganization) {
        Organization organization = scoredOrganization.getOrganization();
        return new OrganizationDto(
            organization.getId().getValue(),
            organization.getName(),
            organization.getUrlName(),
            organization.getDescription(),
            organization.getWebsite(),
            organization.getMapPin(),
            toAddressDtos(organization.getAddresses()),
            toContactPersonDtos(organization.getContactPersons()),
            organization.getLogo(),
            scoredOrganization.getScore()
        );
    }

    static List<OrganizationDto> toOrganizationDtos(List<Organization> organizations) {
        return organizations
            .stream()
            .map(OrganizationAssembler::toOrganizationDto)
            .collect(Collectors.toList());
    }

    private static OrganizationDto toOrganizationDto(Organization organization) {
        return new OrganizationDto(
            organization.getId().getValue(),
            organization.getName(),
            organization.getUrlName(),
            organization.getDescription(),
            organization.getWebsite(),
            organization.getMapPin(),
            toAddressDtos(organization.getAddresses()),
            toContactPersonDtos(organization.getContactPersons()),
            organization.getLogo(),
            null
        );
    }

    static OrganizationDetailDto toOrganizationDetailDto(Organization organization, List<Question> questions) {
        if (organization == null) {
            return null;
        }
        return new OrganizationDetailDto(
            organization.getId().getValue(),
            organization.getName(),
            organization.getUrlName(),
            organization.getOrganizationType(),
            organization.getDescription(),
            organization.getWebsite(),
            organization.getLogo(),
            toPictures(organization.getPictures()),
            toContactPersonDtos(organization.getContactPersons()),
            toAddressDto(organization.getDefaultAddress()),
            toAddressDtos(organization.getAddresses()),
            toAnsweredQuestionDtos(organization.getQuestionAnswers(), questions),
            organization.getMapPin(),
            toGroupDtos(organization.getGroups()),
            toAttendanceTimeDtos(organization.getAttendanceTimes()),
            toVolunteerDtos(organization.getVolunteers())
        );
    }

    private static List<VolunteerDto> toVolunteerDtos(List<Volunteer> volunteers) {
        if (volunteers == null) {
            return Collections.emptyList();
        }
        return volunteers
            .stream()
            .map(OrganizationAssembler::toVolunteerDto)
            .collect(Collectors.toList());
    }

    public static VolunteerDto toVolunteerDto(Volunteer volunteer) {
        return new VolunteerDto(
            volunteer.getFirstname(),
            volunteer.getMotivation(),
            volunteer.getPicture()
        );
    }

    private static List<AttendanceTimeDto> toAttendanceTimeDtos(List<AttendanceTime> attendanceTimes) {
        if (attendanceTimes == null) {
            return Collections.emptyList();
        }
        return attendanceTimes
            .stream()
            .map(OrganizationAssembler::toAttendanceTimeDto)
            .collect(Collectors.toList());
    }

    public static AttendanceTimeDto toAttendanceTimeDto(AttendanceTime attendanceTime) {
        return new AttendanceTimeDto(
            attendanceTime.getDay(),
            attendanceTime.getStart(),
            attendanceTime.getEnd(),
            attendanceTime.getNote(),
            OrganizationAssembler.toGroupDtos(attendanceTime.getGroups())
        );
    }

    private static List<PictureId> toPictures(List<PictureId> pictures) {
        if (pictures == null) {
            return Collections.emptyList();
        }
        return pictures;
    }

    private static List<AnsweredQuestionDto> toAnsweredQuestionDtos(List<QuestionAnswer> questionAnswers, List<Question> questions) {
        if (questionAnswers == null) {
            return Collections.emptyList();
        }
        return questionAnswers.stream()
            .map((question) -> OrganizationAssembler.toAnsweredQuestionDto(question, determineQuestionText(questions, question.getQuestionId())))
            .collect(Collectors.toList());
    }

    public static AnsweredQuestionDto toAnsweredQuestionDto(QuestionAnswer questionAnswer, String question) {
        return new AnsweredQuestionDto(
            questionAnswer.getQuestionId(),
            question,
            questionAnswer.getAnswer()
        );
    }

    public static String determineQuestionText(List<Question> questions, QuestionId id) {
        return questions.stream()
            .filter(question -> Objects.equal(question.getId(), id))
            .map(Question::getQuestion)
            .findFirst()
            .orElse(null);
    }

    private static List<GroupDto> toGroupDtos(List<Group> groups) {
        if (groups == null) {
            return Collections.emptyList();
        }
        return groups.stream().map(OrganizationAssembler::toGroupDto).collect(Collectors.toList());
    }

    public static GroupDto toGroupDto(Group group) {
        return new GroupDto(
            group.getName(),
            group.getDescription()
        );
    }

    private static List<ContactPersonDto> toContactPersonDtos(List<ContactPerson> contactPersons) {
        if (contactPersons == null) {
            return Collections.emptyList();
        }
        return contactPersons
            .stream()
            .map(OrganizationAssembler::toContactPersonDto)
            .collect(Collectors.toList());
    }

    public static ContactPersonDto toContactPersonDto(ContactPerson contactPerson) {
        return new ContactPersonDto(
            contactPerson.getFirstname(),
            contactPerson.getLastname(),
            contactPerson.getRank(),
            contactPerson.getTelephone(),
            contactPerson.getMail(),
            contactPerson.getPicture()
        );
    }

    private static List<AddressDto> toAddressDtos(List<Address> addresses) {
        if (addresses == null) {
            return Collections.emptyList();
        }
        return addresses
            .stream()
            .map(OrganizationAssembler::toAddressDto)
            .collect(Collectors.toList());
    }

    public static AddressDto toAddressDto(Address address) {
        if (address == null) {
            return null;
        }
        return new AddressDto(
            address.getStreet(),
            address.getAddressAppendix(),
            address.getCity(),
            address.getZipcode(),
            address.getLocation(),
            address.getTelephone(),
            address.getWebsite()
        );
    }

    public static Organization toOrganization(OrganizationDetailDto organizationDetailDto) {
        return new Organization.Builder()
            .setId(new OrganizationId(organizationDetailDto.getId()))
            .setName(organizationDetailDto.getName())
            .setUrlName(organizationDetailDto.getUrlName())
            .setDescription(organizationDetailDto.getDescription())
            .setWebsite(organizationDetailDto.getWebsite())
            .setLogo(organizationDetailDto.getLogo())
            .setGroups(toGroups(organizationDetailDto.getGroups()))
            .setPictures(organizationDetailDto.getPictures())
            .setDefaultAddress(toAddress(organizationDetailDto.getDefaultAddress()))
            .setAddresses(toAddresses(organizationDetailDto.getAddresses()))
            .setOrganizationType(organizationDetailDto.getOrganizationType())
            .setAttendanceTimes(toAttendanceTimes(organizationDetailDto.getAttendanceTimes()))
            .setMapPin(organizationDetailDto.getMapPin())
            .setContactPersons(toContactPersons(organizationDetailDto.getContactPersons()))
            .setQuestionAnswers(toQuestionAnswers(organizationDetailDto.getQuestions()))
            .setTeaserImage(toTeaserImage(organizationDetailDto.getPictures()))
            .setVolunteers(toVolunteers(organizationDetailDto.getVolunteers()))
            .build();
    }

    private static List<Volunteer> toVolunteers(List<VolunteerDto> volunteers) {
        if (volunteers == null) {
            return null;
        }
        return volunteers
            .stream()
            .map(OrganizationAssembler::toVolunteer)
            .collect(Collectors.toList());
    }

    public static Volunteer toVolunteer(VolunteerDto volunteerDto) {
        return new Volunteer.Builder()
            .setFirstname(volunteerDto.getFirstname())
            .setMotivation(volunteerDto.getMotivation())
            .setPicture(volunteerDto.getPicture())
            .build();
    }

    private static PictureId toTeaserImage(List<PictureId> pictures) {
        if (pictures == null) {
            return null;
        }
        return pictures
            .stream()
            .findFirst()
            .orElse(null);
    }

    private static List<QuestionAnswer> toQuestionAnswers(List<AnsweredQuestionDto> questions) {
        if (questions == null) {
            return null;
        }
        return questions
            .stream()
            .map(OrganizationAssembler::toQuestionAnswer)
            .collect(Collectors.toList());
    }

    public static QuestionAnswer toQuestionAnswer(AnsweredQuestionDto answeredQuestionDto) {
        return new QuestionAnswer(answeredQuestionDto.getQuestionId(), answeredQuestionDto.getAnswer());
    }

    private static List<ContactPerson> toContactPersons(List<ContactPersonDto> contactPersons) {
        if (contactPersons == null) {
            return null;
        }
        return contactPersons
            .stream()
            .map(OrganizationAssembler::toContactPerson)
            .collect(Collectors.toList());
    }

    public static ContactPerson toContactPerson(ContactPersonDto contactPersonDto) {
        return new ContactPerson.Builder()
            .setFirstname(contactPersonDto.getFirstname())
            .setLastname(contactPersonDto.getLastname())
            .setPicture(contactPersonDto.getPicture())
            .setRank(contactPersonDto.getRank())
            .setTelephone(contactPersonDto.getTelephone())
            .setMail(contactPersonDto.getMail())
            .build();
    }

    private static List<AttendanceTime> toAttendanceTimes(List<AttendanceTimeDto> attendanceTimes) {
        if (attendanceTimes == null) {
            return null;
        }
        return attendanceTimes
            .stream()
            .map(OrganizationAssembler::toAttendanceTime)
            .collect(Collectors.toList());
    }

    public static AttendanceTime toAttendanceTime(AttendanceTimeDto attendanceTimeDto) {
        return new AttendanceTime.Builder()
            .setDay(attendanceTimeDto.getDay())
            .setStart(attendanceTimeDto.getStart())
            .setEnd(attendanceTimeDto.getEnd())
            .setNote(attendanceTimeDto.getNote())
            .setGroups(toGroups(attendanceTimeDto.getGroups()))
            .build();
    }

    private static List<Group> toGroups(List<GroupDto> groups) {
        if (groups == null) {
            return null;
        }
        return groups
            .stream()
            .map(OrganizationAssembler::toGroup)
            .collect(Collectors.toList());
    }

    public static Group toGroup(GroupDto groupDto) {
        return new Group.Builder()
            .setName(groupDto.getName())
            .setDescription(groupDto.getDescription())
            .build();
    }

    private static List<Address> toAddresses(List<AddressDto> addresses) {
        if (addresses == null) {
            return null;
        }
        return addresses
            .stream()
            .map(OrganizationAssembler::toAddress)
            .collect(Collectors.toList());
    }

    public static Address toAddress(AddressDto addressDto) {
        if (addressDto == null) {
            return null;
        }
        return new Address.Builder()
            .setLocation(addressDto.getLocation())
            .setCity(addressDto.getCity())
            .setStreet(addressDto.getStreet())
            .setZipcode(addressDto.getZipcode())
            .setAddressAppendix(addressDto.getAddressAppendix())
            .setTelephone(addressDto.getTelephone())
            .setWebsite(addressDto.getWebsite())
            .build();
    }

}

File: helfomat-api/src/main/kotlin/de/helfenkannjeder/helfomat/api/organization/OrganizationAssembler.kt
package de.helfenkannjeder.helfomat.api.organization

import de.helfenkannjeder.helfomat.api.question.AnsweredQuestionDto
import de.helfenkannjeder.helfomat.core.organization.*
import de.helfenkannjeder.helfomat.core.question.Question
import de.helfenkannjeder.helfomat.core.question.QuestionId

/**
 * @author Valentin Zickner
 */

fun List<ScoredOrganization>.toScoredOrganizationDtos() = this.map { it.toOrganizationDto() }
fun ScoredOrganization.toOrganizationDto(): OrganizationDto {
    val scoredOrganization = this
    val organization = scoredOrganization.organization
    return OrganizationDto(
        organization.id.value,
        organization.name,
        organization.urlName,
        organization.description,
        organization.website,
        organization.defaultAddress?.toAddressDto(),
        organization.addresses.toAddressDtos(),
        organization.contactPersons.toContactPersonDtos(),
        organization.logo,
        scoredOrganization.score,
        organization.organizationType
    )
}

fun List<Organization>.toOrganizationDtos() = this.map { it.toOrganizationDto() }
fun Organization.toOrganizationDto() = OrganizationDto(
    this.id.value,
    this.name,
    this.urlName,
    this.description,
    this.website,
    this.defaultAddress?.toAddressDto(),
    this.addresses.toAddressDtos(),
    this.contactPersons.toContactPersonDtos(),
    this.logo,
    null,
    this.organizationType
)

fun List<Organization>.toOrganizationDetailsDto(questions: List<Question>) = this.map { it.toOrganizationDetailDto(questions) }
fun Organization.toOrganizationDetailDto(questions: List<Question>) = OrganizationDetailDto(
    this.id.value,
    this.name,
    this.urlName,
    this.organizationType,
    this.description,
    this.website,
    this.logo,
    this.pictures,
    this.contactPersons.toContactPersonDtos(),
    this.defaultAddress?.toAddressDto(),
    this.addresses.toAddressDtos(),
    this.questionAnswers.toAnsweredQuestionDtos(questions),
    this.mapPin,
    this.groups.toGroupDtos(),
    this.attendanceTimes.toAttendanceTimeDtos(),
    this.volunteers.toVolunteerDtos()
)

fun List<Volunteer>.toVolunteerDtos() = this.map { it.toVolunteerDto() }
fun Volunteer.toVolunteerDto() = VolunteerDto(this.firstname, this.motivation, this.picture)

fun List<AttendanceTime>.toAttendanceTimeDtos() = this.map { it.toAttendanceTimeDto() }
fun AttendanceTime.toAttendanceTimeDto() = AttendanceTimeDto(this.day, this.start, this.end, this.note, this.groups.toGroupDtos())

fun List<QuestionAnswer>.toAnsweredQuestionDtos(questions: List<Question>) = this.map { it.toAnsweredQuestionDto(questions.getAnswerToQuestion(it.questionId)) }
fun QuestionAnswer.toAnsweredQuestionDto(question: String) = AnsweredQuestionDto(this.questionId, question, this.answer)
fun List<Question>.getAnswerToQuestion(id: QuestionId?) =
    this.filter { question: Question -> question.id == id }
        .map { obj: Question -> obj.question }
        .first()

fun List<Group>.toGroupDtos() = this.map { it.toGroupDto() }
fun Group.toGroupDto() = GroupDto(this.name, this.description)

fun List<ContactPerson>.toContactPersonDtos() = this.map { it.toContactPersonDto() }
fun ContactPerson.toContactPersonDto() = ContactPersonDto(this.firstname, this.lastname, this.rank, this.telephone, this.mail, this.picture)

fun List<Address>.toAddressDtos() = this.map { it.toAddressDto() }
fun Address.toAddressDto(): AddressDto = AddressDto(this.street, this.addressAppendix, this.city, this.zipcode, this.location, this.telephone, this.website)

fun List<OrganizationDetailDto>.toOrganizations(): List<Organization> = this.map { it.toOrganization() }
fun OrganizationDetailDto.toOrganization() = Organization.Builder(
    id = OrganizationId(this.id),
    name = this.name,
    urlName = this.urlName,
    organizationType = this.organizationType
)
    .setDescription(this.description)
    .setWebsite(this.website)
    .setLogo(this.logo)
    .setGroups(this.groups.toGroups())
    .setPictures(this.pictures)
    .setDefaultAddress(this.defaultAddress?.toAddress())
    .setAddresses(this.addresses.toAddresses())
    .setAttendanceTimes(this.attendanceTimes.toAttendanceTimes())
    .setMapPin(this.mapPin)
    .setContactPersons(this.contactPersons.toContactPersons())
    .setQuestionAnswers(this.questions.toQuestionAnswers())
    .setTeaserImage(this.pictures.firstOrNull())
    .setVolunteers(this.volunteers.toVolunteers())
    .build()

fun List<VolunteerDto>.toVolunteers() = this.map { it.toVolunteer() }
fun VolunteerDto.toVolunteer() = Volunteer(firstname = this.firstname, motivation = this.motivation, picture = this.picture)

fun List<AnsweredQuestionDto>.toQuestionAnswers() = this.map { it.toQuestionAnswer() }
fun AnsweredQuestionDto.toQuestionAnswer() = QuestionAnswer(this.questionId, this.answer)

fun List<ContactPersonDto>.toContactPersons() = this.map { it.toContactPerson() }
fun ContactPersonDto.toContactPerson() = ContactPerson(this.firstname, this.lastname, this.rank, this.telephone, this.mail, this.picture)

fun List<AttendanceTimeDto>.toAttendanceTimes() = this.map { it.toAttendanceTime() }
fun AttendanceTimeDto.toAttendanceTime() = AttendanceTime(this.day, this.start, this.end, this.note, this.groups.toGroups())

fun List<GroupDto>.toGroups() = this.map { it.toGroup() }
fun GroupDto.toGroup() = Group(name = this.name, description = this.description)

fun List<AddressDto>.toAddresses() = this.map { it.toAddress() }
fun AddressDto.toAddress() = Address(this.street, this.addressAppendix, this.city, this.zipcode, this.location, this.telephone, this.website)
